File: apps/mobile2/compose/src/commonTest/kotlin/co/typie/ui/component/FontSpecimenTest.kt
package co.typie.ui.component

import androidx.compose.ui.graphics.Color
import co.typie.ui.utils.toHexRgbString
import kotlin.test.Test
import kotlin.test.assertEquals

class FontSpecimenTest {
  @Test
  fun `fontSpecimenUrl includes encoded fallback and color query params`() {
    val result = fontSpecimenUrl(
      fontId = "font-123",
      text = "보통",
      fallbackTexts = listOf("Regular", "400"),
      colorHex = "#F1F1F7",
    )

    assertEquals(
      "${co.typie.Konfig.API_URL}/font/font-123/specimen?text=%EB%B3%B4%ED%86%B5&fallbacks=Regular&fallbacks=400&color=%23F1F1F7",
      result,
    )
  }

  @Test
  fun `fontSpecimenUrl omits blank optional query params`() {
    val result = fontSpecimenUrl(
      fontId = "font-123",
      text = "Bold",
      fallbackTexts = listOf(" ", "Bold"),
      colorHex = null,
    )

    assertEquals(
      "${co.typie.Konfig.API_URL}/font/font-123/specimen?text=Bold",
      result,
    )
  }

  @Test
  fun `familySpecimenFallbacks returns family name only when it differs from display name`() {
    assertEquals(
      listOf("Pretendard"),
      familySpecimenFallbacks(
        displayName = "프리텐다드",
        familyName = "Pretendard",
      ),
    )

    assertEquals(
      emptyList(),
      familySpecimenFallbacks(
        displayName = "Pretendard",
        familyName = "Pretendard",
      ),
    )
  }

  @Test
  fun `weightSpecimenFallbacks keeps secondary text and numeric fallback`() {
    assertEquals(
      listOf("Regular", "400"),
      weightSpecimenFallbacks(
        label = "보통",
        subfamilyDisplayName = "Regular",
        weight = 400,
      ),
    )

    assertEquals(
      listOf("400"),
      weightSpecimenFallbacks(
        label = "보통",
        subfamilyDisplayName = null,
        weight = 400,
      ),
    )

    assertEquals(
      emptyList(),
      weightSpecimenFallbacks(
        label = "400",
        subfamilyDisplayName = null,
        weight = 400,
      ),
    )
  }

  @Test
  fun `toHexRgbString formats opaque colors as uppercase hex`() {
    assertEquals("#F1F1F7", Color(0xFFF1F1F7).toHexRgbString())
  }
}

File: apps/website/src/lib/components/font-specimen.test.ts
import { describe, expect, it } from 'vitest';
import { buildFontSpecimenSearchParams, buildFontSpecimenUrl, familySpecimenFallbacks, weightSpecimenFallbacks } from './font-specimen';

describe('buildFontSpecimenSearchParams', () => {
  it('serializes repeated fallbacks in order', () => {
    expect(buildFontSpecimenSearchParams('보통', ['Regular', '400']).toString()).toBe(
      'text=%EB%B3%B4%ED%86%B5&fallbacks=Regular&fallbacks=400',
    );
  });
});

describe('buildFontSpecimenUrl', () => {
  it('builds the specimen endpoint url with repeated fallbacks', () => {
    expect(buildFontSpecimenUrl('https://api.typie.dev', 'font-123', '보통', ['Regular', '400'])).toBe(
      'https://api.typie.dev/font/font-123/specimen?text=%EB%B3%B4%ED%86%B5&fallbacks=Regular&fallbacks=400',
    );
  });
});

describe('familySpecimenFallbacks', () => {
  it('keeps only the family name when it differs from the display name', () => {
    expect(familySpecimenFallbacks('프리텐다드', 'Pretendard')).toEqual(['Pretendard']);
    expect(familySpecimenFallbacks('Pretendard', 'Pretendard')).toEqual([]);
  });
});

describe('weightSpecimenFallbacks', () => {
  it('keeps subfamily then numeric fallback', () => {
    expect(weightSpecimenFallbacks('보통', 'Regular', 400)).toEqual(['Regular', '400']);
    expect(weightSpecimenFallbacks('400', null, 400)).toEqual([]);
  });
});
